### PR TITLE
Use _RecordFunctionFast for torch.autograd.profiler.record_function

### PR DESCRIPTION
**Pull request by azahed98**
This changes `record_function` to use _RecordFunctionFast, which is more efficient. This is important for Compiler runtime overhead in smaller graphs. Currently this profiler takes half of the overhead.

To measure impact on runtime overhead, here are results from [benchmarks/dynamo/microbenchmarks/overheads.py](https://github.com/pytorch/pytorch/blob/main/benchmarks/dynamo/microbenchmarks/overheads.py).

Before changes:
```
requires_grad=False
compiled 56.9us (warmup=19.3s)

requires_grad=True
compiled 101.6us (warmup=0.2s)

inference_mode()
compiled 53.4us (warmup=0.1s)
```

After changes:
```
requires_grad=False
compiled 27.2us (warmup=9.9s)

requires_grad=True
compiled 61.6us (warmup=0.1s)

inference_mode()
compiled 25.3us (warmup=0.1s)
```

### CODE REVIEW

File: torch/autograd/profiler.py [hunk 1]
```diff
@@ -777,13 +777,11 @@ def __init__(self, name: str, args: Optional[str] = None):
...
         self.record = torch.jit.annotate(
-            Optional["torch.classes.profiler._RecordFunction"], None
+            Optional["torch.classes.profiler._RecordFunctionFast"], None
         )
 
     def __enter__(self):
-        self.record = torch.ops.profiler._record_function_enter_new(
-            self.name, self.args
-        )
+        self.record = torch._C._profiler._RecordFunctionFast(self.name)
```
Comment (sraikund16) [inline torch/autograd/profiler.py]: why are we not adding self.args here?

Comment (sraikund16) [reply torch/autograd/profiler.py]: also you will need to make the scope USER as done in this PR https://github.com/pytorch/pytorch/pull/162661

**Comment by sraikund16** [comment]
Please also add the trace itself to make sure that it is 1:1 with the same workload before this change

File: torch/autograd/profiler.py (hunk 1 above)
Comment (azahed98) [reply torch/autograd/profiler.py]: Ah apologies I was using and outdated signature from here https://github.com/pytorch/pytorch/pull/107195/files#diff-1992ddca5234f6df084fdcaf2fe86216c2ab2b58f21300517284d163b579ca88

Passing self.args results in an abort. Should it instead be `torch._C._profiler._RecordFunctionFast(self.name, (self.args, ))`?

**Comment by azahed98** [comment]
@sraikund16 Traceback comparison using the example from the comment under `record_function`.

Without change (baseline):
```
-------------------------------------------------------  ------------  ------------  ------------  ------------  ------------  ------------  
                                                   Name    Self CPU %      Self CPU   CPU total %     CPU total  CPU time avg    # of Calls  
-------------------------------------------------------  ------------  ------------  ------------  ------------  ------------  ------------  
                                              aten::pow        30.11%     289.843us        38.51%     370.719us     123.573us             3  
                                                label-z        19.17%     184.546us        22.47%     216.278us     216.278us             1  
                                              aten::mul         7.83%      75.415us        15.07%     145.054us      72.527us             2  
                                            aten::copy_         6.47%      62.266us         6.47%      62.266us      31.133us             2  
                                           PowBackward0         6.20%      59.679us        31.52%     303.380us     303.380us             1  
                                        aten::ones_like         5.31%      51.145us         9.70%      93.387us      93.387us             1  
      autograd::engine::evaluate_function: PowBackward0         4.42%      42.538us        35.94%     345.918us     345.918us             1  
                        torch::autograd::AccumulateGrad         3.33%      32.017us         6.44%      62.009us      62.009us             1  
                                         aten::_to_copy         3.03%      29.180us         5.30%      51.006us      51.006us             1  
                                               aten::to         2.93%      28.181us         8.23%      79.187us      19.797us             4  
                                                 detach         2.79%      26.864us         2.79%      26.864us      26.864us             1  
                                      aten::result_type         2.61%      25.099us         2.61%      25.099us       8.366us             3  
                                    aten::empty_strided         2.07%      19.928us         2.07%      19.928us       9.964us             2  
                                       aten::empty_like         1.77%      16.997us         3.23%      31.136us      31.136us             1  
                                            aten::fill_         1.15%      11.106us         1.15%      11.106us      11.106us             1  
autograd::engine::evaluate_function: torch::autograd...         0.48%       4.626us         6.92%      66.635us      66.635us             1  
                                           aten::detach         0.32%       3.128us         3.12%      29.992us      29.992us             1  
-------------------------------------------------------  ------------  ------------  ------------  ------------  ------------  ------------  
Self CPU time total: 962.558us
```

With change:
```
-------------------------------------------------------  ------------  ------------  ------------  ------------  ------------  ------------  
                                                   Name    Self CPU %      Self CPU   CPU total %     CPU total  CPU time avg    # of Calls  
-------------------------------------------------------  ------------  ------------  ------------  ------------  ------------  ------------  
                                              aten::pow        38.54%     306.282us        48.69%     386.945us     128.982us             3  
                                              aten::mul         8.49%      67.493us        18.89%     150.097us      75.048us             2  
                                            aten::copy_         7.99%      63.528us         7.99%      63.528us      31.764us             2  
                                           PowBackward0         6.88%      54.646us        38.47%     305.712us     305.712us             1  
                                               aten::to         5.17%      41.058us        11.53%      91.673us      22.918us             4  
                                        aten::ones_like         5.02%      39.916us        12.01%      95.411us      95.411us             1  
                                    aten::empty_strided         4.25%      33.769us         4.25%      33.769us      16.884us             2  
                        torch::autograd::AccumulateGrad         3.77%      29.992us         7.28%      57.862us      57.862us             1  
                                         aten::_to_copy         3.67%      29.133us         6.37%      50.615us      50.615us             1  
      autograd::engine::evaluate_function: PowBackward0         3.32%      26.425us        41.79%     332.137us     332.137us             1  
                                      aten::result_type         3.16%      25.089us         3.16%      25.089us       8.363us             3  
                                                 detach         3.14%      24.986us         3.14%      24.986us      24.986us             1  
                                                label-z         2.35%      18.700us         6.06%      48.189us      48.189us             1  
                                       aten::empty_like         1.97%      15.652us         5.66%      44.962us      44.962us             1  
                                            aten::fill_         1.33%      10.533us         1.33%      10.533us      10.533us             1  
autograd::engine::evaluate_function: torch::autograd...         0.59%       4.672us         7.87%      62.534us      62.534us             1  
                                           aten::detach         0.36%       2.884us         3.51%      27.870us      27.870us             1  
-------------------------------------------------------  ------------  ------------  ------------  ------------  ------------  ------------  
Self CPU time total: 794.758us
```

I see that `label-z` significantly drops. Is this just from reduced CPU time in moving to _RecordFunctionFast?

**Comment by azahed98** [comment]
Scope of this PR is larger than I initially anticipated, and would likely require discussions with others about `record_function` usage and if `_RecordFunction` should be deprecated following this PR's intended change. Closing for now, to be revisited later.
